packages/neptune-extension/src/common/components/UploadForm.tsx

```diff
@@ -0,0 +1,164 @@
+import React from 'react';
+
+import { leaderboardClient } from 'common/api/leaderboard-client';
+import { NotebookDTO } from 'generated/leaderboard-client/src/models';
+
+import {
+  PlatformNotebook,
+  PlatformNotebookMetadata,
+} from 'types/platform';
+
+import Modal from 'common/components/Modal';
+
+interface UploadFormProps {
+  platformNotebook: PlatformNotebook
+  notebook: NotebookDTO | undefined
+  defaultProjectId: string
+  onCancel: () => void
+  onSuccess: () => void
+}
+
+const UploadForm: React.FC<UploadFormProps> = ({
+  platformNotebook,
+  notebook,
+  defaultProjectId,
+  onCancel,
+  onSuccess,
+}) => {
+  const [ metadata ] = React.useState(() => platformNotebook.getMetadata());
+
+  const canUploadCheckpoint = notebook
+    && notebook.id === metadata.notebookId;
+    // TODO: check the owner of notebook
+
+  const [ loading, setLoading ] = React.useState(false);
+
+  const [ mode, setMode ] = React.useState(canUploadCheckpoint ? 'checkpoint' : 'notebook');
+
+  const [ projectId, setProjectId ] = React.useState(() => {
+    return !!notebook
+      ? notebook.projectId
+      : defaultProjectId;
+  });
+  const [ name, setName ] = React.useState('');
+  const [ description, setDescription ] = React.useState('');
+
+  function changeMode(newMode: string) {
+    if (newMode === 'checkpoint') {
+      setProjectId((notebook as NotebookDTO).projectId);
```

Comment:
I think it is not good practice to cast the type of this object if it is optional.
There is of course in current code no possibility that notebook will be undefined but this assumtion is very implicit.

If someone will change the code of `canUploadCheckpoint` this logic will break.

I would at least mark it by the comment to explain why it is casted.
But to be honest a simple if statement, checking if it is not undefined would be just more readable.